lighthouse-core/audits/resource-budget.js

```diff
@@ -0,0 +1,185 @@
+  }
+
+  /**
+* @param {{resourceType: LH.Budget.ResourceType, count: number, size: number}} row
+* @param {LH.Budget | undefined} budget
+* @return {boolean}
+*/
+  static shouldIncludeRow(row, budget) {
+    if (budget === undefined) {
+      return true;
+    } else {
+      const budgets = (budget.resourceCounts || []).concat(budget.resourceSizes || []);
+      return !!budgets.find((b) => {
+        return b.resourceType === row.resourceType;
+      });
+    }
+  }
+
+  /**
+* @param {LH.Budget|undefined} budget
+* @param {Object<string,{resourceType: LH.Budget.ResourceType, count: number, size: number}>} summary
+* @return {Array<{label: string, count: number, size: number, countOverBudget?: string|undefined, sizeOverBudget?: number|undefined}>}
+*/
+  static tableItems(budget, summary) {
+    /** @type {Object<LH.Budget.ResourceType,string>} */
+    const strMappings = {
+      'total': str_(UIStrings.totalResourceType),
+      'document': str_(UIStrings.documentResourceType),
+      'script': str_(UIStrings.scriptResourceType),
+      'stylesheet': str_(UIStrings.stylesheetResourceType),
+      'image': str_(UIStrings.imageResourceType),
+      'media': str_(UIStrings.mediaResourceType),
+      'font': str_(UIStrings.fontResourceType),
+      'other': str_(UIStrings.otherResourceType),
+      'third-party': str_(UIStrings.thirdPartyResourceType),
+    };
+
+    /** @type {Array<{label: string, count: number, size: number, countOverBudget?: string|undefined, sizeOverBudget?: number|undefined}>}*/
+    return Object.values(summary).filter((row) => {
+      return this.shouldIncludeRow(row, budget);
+    }).map((row) => {
+      const type = row.resourceType;
+
+      const overCount = this.overBudgetAmount(budget && budget.resourceCounts, row.count, type);
+      const overSize = this.overBudgetAmount(budget && budget.resourceSizes, row.size, type);
+
+      return {
+        label: strMappings[type],
+        count: row.count,
+        size: row.size,
+        countOverBudget: overCount !== undefined ?
+          str_(UIStrings.requestCount, {count: overCount}) : undefined,
+        sizeOverBudget: overSize,
+      };
+    }).sort((a, b) => {
+      const overBudgetComparison = (b.sizeOverBudget || 0) - (a.sizeOverBudget || 0);
+      const sizeComparison = b.size - a.size;
+      return budget ? overBudgetComparison : sizeComparison;
+    });
+  }
+
+  /**
+     * @param {LH.Artifacts} artifacts
+     * @param {LH.Audit.Context} context
+     * @return {Promise<LH.Audit.Product>}
+     */
+  static async audit(artifacts, context) {
+    const devtoolsLog = artifacts.devtoolsLogs[Audit.DEFAULT_PASS];
+    const networkRecords = await NetworkRecords.request(devtoolsLog, context);
+    const mainResource = await (MainResource.request({devtoolsLog, URL: artifacts.URL}, context));
+
+    /** @type{LH.Budget | undefined} */
+    const budget = (context.settings.budgets || []).reverse().find((budget) => {
```

Comment:
early returns when there's no budgets might make this clearer. Could we have a different function called/execution path for that case rather than interleaving the two and having to handle the `undefined` case all over the place?